### PR TITLE
Added formatter for bit_reference-like types

### PR DESCRIPTION
**Pull request by phprus**
Fix for issue #3567

Test is based on #3569

### CODE REVIEW

**Comment by phprus** [comment]
Final version.
Ready for review.

File: include/fmt/std.h [hunk 1]
```diff
@@ -389,6 +390,16 @@ struct formatter<
 #endif
   }
 };
+
+FMT_EXPORT
+template <typename Char>
+struct formatter<std::vector<bool>::reference, Char> : formatter<bool, Char> {
```
Comment (wangzw) [inline include/fmt/std.h:396]: Maybe need to handle non default allocator

Comment (phprus) [reply include/fmt/std.h:396]: We can't deduce Allocator type in this context.

```
/Users/phprus/Devel/fmt/include/fmt/std.h:395:8: error: class template partial specialization contains a template parameter that cannot be deduced; this partial specialization will never be used [-Wunusable-partial-specialization]
struct formatter<typename std::vector<bool, Allocator>::reference, Char>
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/phprus/Devel/fmt/include/fmt/std.h:394:20: note: non-deducible template parameter 'Allocator'
template <typename Allocator, typename Char>
                   ^
```

**Comment by phprus** [comment]
@vitaut

Final-2 version :)
Ready for review.

File: include/fmt/std.h [hunk 2]
```diff
@@ -389,6 +391,38 @@ struct formatter<
...
+template <typename T>
+struct has_flip<T, void_t<decltype(std::declval<T>().flip())>>
+    : std::true_type {};
+
+template <typename T> struct is_bit_reference_like {
+  static constexpr const bool value =
+      std::is_convertible<T, bool>::value &&
+      std::is_nothrow_assignable<T, bool>::value && has_flip<T>::value
+
+      ;
```
Comment (wangzw) [inline include/fmt/std.h:409]: formatter issue?

Comment (phprus) [reply include/fmt/std.h]: Unexpected clang-format behaviour.
Fixed.

File: test/std-test.cc [hunk 3]
```diff
@@ -221,3 +222,10 @@ TEST(std_test, exception) {
   }
 #endif
 }
+
+TEST(std_test, format_bit_reference) {
+  std::bitset<2> bs(1);
+  EXPECT_EQ(fmt::format("{} {}", bs[0], bs[1]), "true false");
+  std::vector<bool> v = {true, false};
+  EXPECT_EQ(fmt::format("{} {}", v[0], v[1]), "true false");
```
Comment (wangzw) [inline test/std-test.cc:230]: try this

```
  const std::vector<bool> cv = {true, false};
  EXPECT_EQ(fmt::format("{} {}", cv[0], cv[1]), "true false");
```

```
[ 90%] Building CXX object test/CMakeFiles/std-test.dir/std-test.cc.o
In file included from /Users/wangzw/workspace/fmt/test/std-test.cc:8:
In file included from /Users/wangzw/workspace/fmt/include/fmt/std.h:21:
In file included from /Users/wangzw/workspace/fmt/include/fmt/format.h:49:
/Users/wangzw/workspace/fmt/include/fmt/core.h:1621:63: error: implicit instantiation of undefined template 'fmt::detail::type_is_unformattable_for<std::__bit_const_reference<std::vector<bool>>, char>'
    type_is_unformattable_for<T, typename Context::char_type> _;
                                                              ^
/Users/wangzw/workspace/fmt/include/fmt/core.h:1853:23: note: in instantiation of function template specialization 'fmt::detail::make_arg<true, fmt::basic_format_context<fmt::appender, char>, std::__bit_const_reference<std::vector<bool>>, 0>' requested here
        data_{detail::make_arg<is_packed, Context>(args)...} {
                      ^
/Users/wangzw/workspace/fmt/include/fmt/core.h:1871:10: note: in instantiation of function template specialization 'fmt::format_arg_store<fmt::basic_format_context<fmt::appender, char>, std::__bit_const_reference<std::vector<bool>>, std::__bit_const_reference<std::vector<bool>>>::format_arg_store<std::__bit_const_reference<std::vector<bool>>, std::__bit_const_reference<std::vector<bool>>>' requested here
  return {args...};
         ^
/Users/wangzw/workspace/fmt/include/fmt/core.h:2833:28: note: in instantiation of function template specialization 'fmt::make_format_args<fmt::basic_format_context<fmt::appender, char>, std::__bit_const_reference<std::vector<bool>>, std::__bit_const_reference<std::vector<bool>>>' requested here
  return vformat(fmt, fmt::make_format_args(args...));
                           ^
/Users/wangzw/workspace/fmt/test/std-test.cc:231:18: note: in instantiation of function template specialization 'fmt::format<std::__bit_const_reference<std::vector<bool>>, std::__bit_const_reference<std::vector<bool>>>' requested here
  EXPECT_EQ(fmt::format("{} {}", cv[0], cv[1]), "true false");
                 ^
/Users/wangzw/workspace/fmt/include/fmt/core.h:1599:45: note: template is declared here
template <typename T, typename Char> struct type_is_unformattable_for;
                                            ^
/Users/wangzw/workspace/fmt/include/fmt/core.h:1624:3: error: static assertion failed due to requirement 'formattable': Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt
  static_assert(
  ^
```

Comment (phprus) [reply test/std-test.cc:230]: Fixed

**Review by vitaut — CHANGES_REQUESTED** [review]
Thanks for the PR.

This looks like a lot of effort and implementation-specific code for such a corner case. I suggest just using std::addressoff and replacing FMT_CONSTEXPR with FMT_COSTEXPR20 in the value ctor overload that uses it. It only affects compile-time formatting anyway that requires a recent standard.

**Comment by phprus** [comment]
@vitaut 
Done.

**Comment by wangzw** [comment]
Works well for me. Thanks

File: include/fmt/std.h [hunk 4]
```diff
@@ -389,6 +391,52 @@ struct formatter<
...
+      std::is_convertible<T, bool>::value &&
+      std::is_nothrow_assignable<T, bool>::value && has_flip<T>::value;
+};
+
+}  // namespace detail
+
+FMT_EXPORT
+template <typename BitRef, typename Char>
+struct formatter<BitRef, Char,
+                 enable_if_t<detail::is_bit_reference_like<BitRef>::value>>
```
Comment (vitaut) [inline include/fmt/std.h:415]: I don't think we need SFINAE here. Let's use `std::vector<bool>::reference`.

File: include/fmt/std.h [hunk 5]
```diff
@@ -389,6 +391,52 @@ struct formatter<
...
+  }
+};
+
+#ifdef _LIBCPP_VERSION
+
+// Workaround for libc++ incompatibility with C++ standard.
+// According to the Standard, `bitset::operator[] const` returns bool.
+FMT_EXPORT
+template <typename C, typename Char>
+struct formatter<std::__bit_const_reference<C>, Char> : formatter<bool, Char> {
```
Comment (vitaut) [inline include/fmt/std.h:430]: Similarly, this should be `bitset::reference` and remove ifdef and a comment.

Comment (phprus) [reply include/fmt/std.h:430]: See: https://github.com/fmtlib/fmt/pull/3570#discussion_r1285057473

Comment (vitaut) [reply include/fmt/std.h:430]: I suggest moving this to a separate PR and adding more context why this is needed.

File: include/fmt/std.h (hunk 4 above)
Comment (phprus) [reply include/fmt/std.h:415]: 1) The compiler won't be deduce the allocator type from ``std::vector<bool, Allocator>::reference`` or ``N`` value from ``std::bitset<N>::reference``.

2) What is ``std::vector<bool>::reference`` or ``std::bitset<N>::reference``?
On libc++ is ``std::__bit_reference`` and ``std::__bit_const_reference``.
On gcc and msvs ``std::vector<bool>::reference`` and ``std::bitset<N>::reference`` - is different proxy classes.

We need separate implementation for libc++ and gcc/msvc. But in this case, we will not be able to implement output for ``std::bitset``.

See:
https://github.com/fmtlib/fmt/pull/3570#discussion_r1284995184 (Allocator)
https://github.com/fmtlib/fmt/pull/3570#discussion_r1285057473 (const)

Previous branch:
https://github.com/fmtlib/fmt/tree/85bc49b83d59070a8023f529ec80c38652c14910

Commit with formatters: https://github.com/fmtlib/fmt/commit/64c905e27868c246289fd5d5e13663cc7410cfce
Failed tests: https://github.com/fmtlib/fmt/actions/runs/5769810223/job/15642249348 , https://github.com/fmtlib/fmt/actions/runs/5769810222/job/15642249612

File: include/fmt/std.h (hunk 5 above)
Comment (phprus) [reply include/fmt/std.h:430]: > I suggest moving this to a separate PR and adding more context why this is needed.

I do not understand why.
Without this workaround we can't format values from ``const`` containers on libc++.

File: include/fmt/std.h (hunk 4 above)
Comment (vitaut) [reply include/fmt/std.h:415]: Why do we care about allocator? Reference type should normally be the same between vectors with different allocators: https://godbolt.org/z/5sKM5b6vb.

Comment (phprus) [reply include/fmt/std.h:415]: No. It is implementation specific behaviour:
libc++: https://godbolt.org/z/3xbxj89hn

Comment (vitaut) [reply include/fmt/std.h:415]: That's unfortunate.

File: include/fmt/std.h [hunk 6]
```diff
@@ -389,6 +391,52 @@ struct formatter<
...
+
+FMT_EXPORT
+template <typename BitRef, typename Char>
+struct formatter<BitRef, Char,
+                 enable_if_t<detail::is_bit_reference_like<BitRef>::value>>
+    : formatter<bool, Char> {
+  template <typename FormatContext>
+  FMT_CONSTEXPR auto format(const BitRef& v, FormatContext& ctx) const
+      -> decltype(ctx.out()) {
+    return formatter<bool, Char>::format(static_cast<bool>(v), ctx);
```
Comment (vitaut) [inline include/fmt/std.h:420]: I don't think this cast is needed because the reference should be implicitly convertible to bool.

File: include/fmt/std.h (hunk 5 above)
Comment (vitaut) [reply include/fmt/std.h:430]: It would make sense to move if we used concrete types but in view of the above discussion it seems impossible so ignore this comment.

File: include/fmt/std.h [hunk 7]
```diff
@@ -389,6 +391,52 @@ struct formatter<
...
+struct has_flip : std::false_type {};
+
+template <typename T>
+struct has_flip<T, void_t<decltype(std::declval<T>().flip())>>
+    : std::true_type {};
+
+template <typename T> struct is_bit_reference_like {
+  static constexpr const bool value =
+      std::is_convertible<T, bool>::value &&
+      std::is_nothrow_assignable<T, bool>::value && has_flip<T>::value;
```
Comment (vitaut) [inline include/fmt/std.h:407]: Can we remove the workaround for libc++ if we drop the `is_nothrow_assignable` check?

File: include/fmt/std.h (hunk 4 above)
Comment (vitaut) [reply include/fmt/std.h:415]: A short comment explaining why we can't use `vector<bool>::reference` would be nice though.

File: include/fmt/std.h (hunk 7 above)
Comment (phprus) [reply include/fmt/std.h:407]: No. ``__bit_const_reference`` does not contain ``flip`` method.

``is_bit_reference_like`` - is standard compatibility reference.
``__bit_const_reference`` is libc++ non standard issue.

File: include/fmt/std.h (hunk 4 above)
Comment (phprus) [reply include/fmt/std.h:418]: Done.

File: include/fmt/std.h (hunk 6 above)
Comment (phprus) [reply include/fmt/std.h]: Done.

File: include/fmt/std.h [hunk 8]
```diff
@@ -389,6 +391,52 @@ struct formatter<
...
+template <typename C, typename Char>
+struct formatter<std::__bit_const_reference<C>, Char> : formatter<bool, Char> {
+  template <typename FormatContext>
+  FMT_CONSTEXPR20 auto format(const std::__bit_const_reference<C>& v,
+                              FormatContext& ctx) const -> decltype(ctx.out()) {
+    return formatter<bool, Char>::format(v, ctx);
+  }
+};
+
+#endif
```
Comment (vitaut) [inline include/fmt/std.h:441]: I think this can be replaced by a specialization of `is_bit_reference_like` for `__bit_const_reference` which seems a bit cleaner (no duplicate formatter).

Comment (phprus) [reply include/fmt/std.h]: Done.

**Review by vitaut — APPROVED** [review]

**Comment by vitaut** [comment]
Thank you!
